spec.html

```diff
@@ -4196,6 +4196,43 @@ <h1>UpdateEmpty ( _completionRecord_, _value_ )</h1>
           1. Return Completion { [[Type]]: _completionRecord_.[[Type]], [[Value]]: _value_, [[Target]]: _completionRecord_.[[Target]] }.
         </emu-alg>
       </emu-clause>
+
+      <emu-clause id="sec-yield" aoid="Yield">
+        <h1>Yield</h1>
+
+        <p>Algorithm steps that say</p>
+
+        <emu-alg>
+          1. Let _completion_ be Yield(_value_).
+        </emu-alg>
+
+        <p>mean the same thing as:</p>
+        <emu-alg>
+          1. Let _generator_ be the Function component of the running execution context.
+          1. Perform ? GeneratorYield(CreateIterResultObject(_value, *false*)).
+          1. Assert: The execution context of _generator_ is suspended.
+          1. When the running execution context is resumed by GeneratorResume(_generator_, _yieldNextValue_, _generatorBrand_) at step <emu-xref href="#step-generator-resume-resume-execution-context"></emu-xref>, let _completion_ be NormalCompletion(_yieldNextValue_).
+          1. When the running execution context is resumed by GeneratorResumeAbrupt(_runningContext_, _abruptCompletion_, _generatorBrand_) at step <emu-xref href="#step-generator-abrupt-resume-execution-context"></emu-xref>, let _completion_ be _abruptCompletion_.
+          1. If the _generator_.[[GeneratorState]] is not ~executing~, return NormalCompletion(*undefined*).
+          1. Assert: _completion_ is a normal completion or an abrupt completion.
```

Comment:
I don't think this needs to be a macro, you just need to copy the steps of |YieldExpression| into an abstract op:

### Yield ( _value_ )
1. Let _generatorKind_ be ! GetGeneratorKind().
1. If _generatorKind_ is async, then return ? AsyncGeneratorYield(_value_).
1. Otherwise, return ? GeneratorYield(CreateIterResultObject(_value_, false)).

### YieldExpression : yield
1. Return ? Yield(undefined).

### YieldExpression : yield *AssignmentExpression*
1. Let _exprRef_ be the result of evaluating *AssignmentExpression*.
1. Let _value_ be ? GetValue(_exprRef_).
1. Return ? Yield(_value_).

Reply:
But this does match how [**Await**](https://tc39.es/ecma262/#await) is defined.

Reply:
Yes, I'm following the way of Await op

Reply:
But the current way seems like not supporting async iterator 🤔

Reply:
But the current way seems like not supporting async iterator 🤔

Reply:
the generator yield functions are already abstract ops instead of macros. the only thing that needs to happen here is moving the logic in `yield` to something that builtin generators can reuse. like you could just directly call GeneratorYield() or AsyncGeneratorYield() in a builtin generator but it seems annoying to keep track of that.